### PR TITLE
Fix intermittent CI failure in EmptyQueue

### LINKED ISSUE #23608: Caught a flaky test:
Caught a flaky test:

https://drone.gitea.io/go-gitea/gitea/70256/2/14

```
--- FAIL: TestPersistableChannelUniqueQueue (0.62s)

    unique_queue_disk_channel_test.go:227: TestPersistableChannelUniqueQueue executedInitiallyA=57, executedInitiallyB=75, executedToEmptyA=44, executedToEmptyB=26

    --- FAIL: TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01 (0.17s)

        unique_queue_disk_channel_test.go:210: 

            	Error Trace:	/drone/src/modules/queue/unique_queue_disk_channel_test.go:210

            	Error:      	Not equal: 

            	            	expected: 101

            	            	actual  : 100

            	Test:       	TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01

        unique_queue_disk_channel_test.go:211: 

            	Error Trace:	/drone/src/modules/queue/unique_queue_disk_channel_test.go:211

            	Error:      	Not equal: 

            	            	expected: 0

            	            	actual  : 1

            	Test:       	TestPersistableChannelUniqueQueue/Empty_Queue:_QueueA#01

    unique_queue_disk_channel_test.go:256: TestPersistableChannelUniqueQueue executedInitiallyA=21, executedInitiallyB=100, executedToEmptyA=79, executedToEmptyB=1

FAIL
```

_Originally posted by @wxiaoguang in https://github.com/go-gitea/gitea/issues/23154#issuecomment-1477268999_

### LINKED ISSUE #23977: Intermittent CI Failure TestPersistableChannelUniqueQueue
### Description

```log
ok  	code.gitea.io/gitea/modules/public	0.100s	coverage: 7.7% of statements

2023/04/07 10:45:59 ...ueue_channel_test.go:169:func3() [I] Finally terminating

2023/04/07 10:45:59 ...eue/queue_channel.go:129:func1() [W] ChannelQueue: test-queue-channel Terminated before completed flushing

2023/04/07 10:46:00 ...disk_channel_test.go:202:func1() [I] [642ff468-2] pausing

2023/04/07 10:46:01 ...disk_channel_test.go:202:func1() [I] [642ff468-4] pausing

2023/04/07 10:46:01 ...disk_channel_test.go:257:func3() [I] Finally terminating

TempDir /tmp/TestPersistableChannelUniqueQueue1161295328/001

2023/04/07 10:46:03 ...ueue_disk_channel.go:314:Shutdown() [W] [642ff46a-7] PersistableChannelUniqueQueue: QueueB 79 will be restored on restart

panic: send on closed channel


goroutine 361 [running]:

code.gitea.io/gitea/modules/queue.(*WorkerPool).pushBoost(0xc0008c0750, {0x103fd60, 0xc00023dbe0})

	/drone/src/modules/queue/workerpool.go:153 +0xb6

code.gitea.io/gitea/modules/queue.(*WorkerPool).Push(0xc0008c0750, {0x103fd60, 0xc00023dbe0})

	/drone/src/modules/queue/workerpool.go:108 +0x1f9

code.gitea.io/gitea/modules/queue.(*ChannelUniqueQueue).PushFunc(0xc000329340, {0x103fd60, 0xc00023dbe0}, 0x0)

	/drone/src/modules/queue/unique_queue_channel.go:143 +0x2bd

code.gitea.io/gitea/modules/queue.(*PersistableChannelUniqueQueue).PushFunc(0xc0003292d0, {0x103fd60, 0xc00023dbe0}, 0x48a05b?)

	/drone/src/modules/queue/unique_queue_disk_channel.go:153 +0x130

code.gitea.io/gitea/modules/queue.(*PersistableChannelUniqueQueue).Push(0xc002205baf?, {0x103fd60, 0xc00023dbe0})

	/drone/src/modules/queue/unique_queue_disk_channel.go:144 +0x54

code.gitea.io/gitea/modules/queue.NewPersistableChannelUniqueQueue.func2({0xc000272e50, 0x1, 0x10?})

	/drone/src/modules/queue/unique_queue_disk_channel.go:105 +0xf9

code.gitea.io/gitea/modules/queue.NewByteFIFOUniqueQueue.func1({0xc000272e50, 0x1, 0x1})

	/drone/src/modules/queue/queue_bytefifo.go:398 +0x93

code.gitea.io/gitea/modules/queue.(*WorkerPool).FlushWithContext(0xc0002b0f30, {0x25137c8, 0xc000276140})

	/drone/src/modules/queue/workerpool.go:503 +0x4c9

code.gitea.io/gitea/modules/queue.(*WorkerPool).Flush(0x4889ac?, 0x48a05b?)

	/drone/src/modules/queue/workerpool.go:458 +0xaf

code.gitea.io/gitea/modules/queue.(*ByteFIFOQueue).Flush(0xc0002be000, 0x396c120?)

	/drone/src/modules/queue/queue_bytefifo.go:150 +0xc5

code.gitea.io/gitea/modules/queue.(*PersistableChannelUniqueQueue).Run.func4()

	/drone/src/modules/queue/unique_queue_disk_channel.go:219 +0xdf

created by code.gitea.io/gitea/modules/queue.(*PersistableChannelUniqueQueue).Run

	/drone/src/modules/queue/unique_queue_disk_channel.go:216 +0x6f2

FAIL	code.gitea.io/gitea/modules/queue	3.668s
```

### Gitea Version

main

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

run the CI

### Database

None

### PR DESCRIPTION
**Pull request by zeripath**
The ordering of the final token causing a close of the queue in this test may be out of sync due to concurrency. Instead just use ensure that the queue is closed when everything expected is done.

Fixes: https://github.com/go-gitea/gitea/issues/23608
Fixes: https://github.com/go-gitea/gitea/issues/23977

### CODE REVIEW

**Comment by KN4CK3R** [comment]
CI fail is related

File: modules/queue/unique_queue_disk_channel_test.go [hunk 1]
```diff
@@ -154,33 +155,36 @@ func TestPersistableChannelUniqueQueue(t *testing.T) {
 			assert.Equal(t, 101, len(executedInitial[name])+len(hasInitial[name]))
 			mapLock.Unlock()
 		})
+		done <- len(hasInitial[name])
```
Comment (KN4CK3R) [inline modules/queue/unique_queue_disk_channel_test.go]:
```suggestion
		mapLock.Lock()
		done <- len(hasInitial[name])
		mapLock.Unlock()
```
Or assign `len(hasInitial[name])` to a variable first. I don't know if locking around a channel may have side effects.

**Comment by zeripath** [comment]
Sorry about the race - I forgot that there were two things that were simultaneously writing to that map so assumed it didn't need a lock

**Comment by silverwind** [comment]
Still failing it seems.

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/24419 had disabled two queue tests on CI, if they are stable again, remove the disablement.

**Comment by wxiaoguang** [comment]
The old queue code is unmaintainable now, I have starting the rewriting. I believe the queue package can be hugely simplified.

**Comment by zeripath** [comment]
The problem in this PR was simple; the atomic int needed to be defined outside of the handler and the count needs to be pushed to the done channel outside of the lock.

@wxiaoguang the queue code complexity is a natural result of the requirements. You're just going to create more bugs and reiterate errors.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Added issue references to OP.

**Review by techknowlogick — APPROVED** [review]

**Comment by wxiaoguang** [comment]
> @wxiaoguang the queue code complexity is a natural result of the requirements. You're just going to create more bugs and reiterate errors.

Nope, I never created more bugs. I will finish the rewriting soon.

**Comment by wxiaoguang** [comment]
> The problem in this PR was simple; the atomic int needed to be defined outside of the handler and the count needs to be pushed to the done channel outside of the lock.
> 
> @wxiaoguang the queue code complexity is a natural result of the requirements. You're just going to create more bugs and reiterate errors.

@zeripath Just happen to see this un-replied comment, and the new queue system has been in production for more than 1 years, so reply it:  the new code is right, it only uses less than a half of lines (reduces thousands of lines), and it is incredibly stable and no bug ever happens. All users are quite happy.

**Pull request merged**